stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,143 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+
+struct Boxed[T: AnyType, address_space: AddressSpace = AddressSpace.GENERIC]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist. Consider Boxed[T] over UnsafePointer[T] where possible.
+
+    Parameters:
+        T: The type to be stored in the Boxed[].
+        address_space: The address space that the allocation behind this box resides in.
+    """
+
+    var _inner: UnsafePointer[T, address_space]
+
+    fn __init__[
+        T: Movable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
+        """Construct a new Boxed[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, AddressSpace.GENERIC].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_value: T):
+        """Construct a new Boxed[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, address_space].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](
+        inout self: Boxed[T, AddressSpace.GENERIC],
+        *,
+        copy_box: Boxed[T, AddressSpace.GENERIC],
+    ):
+        """Construct a new Boxed[] by explicitly copying the value from another Boxed[].
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_box: The Boxed[] to copy.
+        """
+        self.__init__(copy_value=copy_box[])
+
```

Comment:
**Suggestion** Let's also add a constructor for `T: Copyable` (i.e. when `T` is implicitly copyable).  Right now `ExplicitCopyable` is in a half state (and a lot of the stdlib types aren't there yet), so to make `Box` useful out of the box (no pun intended 😆 ), let's add this additional constructor.

Reply:
I'm not quite sure how I'd want to disambiguate the copying from the moving `__init__`, as I want people to reach for the moving `__init__` "by default" but also don't want the implicit copying constructor to require a whole song and dance just to call it.`Box(copy_value=foo)` is currently the explicit one, I could change that to `Box(explicit_copy_value=foo)` but that verges on mocking the user for daring to not `move`. Is there prior work for this that we'd like to conform to?